Fix #74922 - Try to resolve constants when importing trait properties

pmmaga opened the pull request:
Link for bugsnet: https://bugs.php.net/bug.php?id=74922

It fixes [the reported issue](https://3v4l.org/kkHFe):
```
const VALUE = true;

trait Foo {public $var = VALUE;}
trait Bar {public $var = VALUE;}
class Baz {use Foo, Bar;}
```
And it also allows for values that are identical to the constant like:
```
const VALUE = true;

trait Foo {public $var = VALUE;}
trait Bar {public $var = true;}
class Baz {use Foo, Bar;}
```

nikic commented on Zend/zend_inheritance.c:
```diff
@@ -1565,19 +1566,32 @@ static void zend_do_traits_property_binding(zend_class_entry *ce) /* {{{ */
+						}
+						if (UNEXPECTED(Z_TYPE_P(op2) == IS_CONSTANT)) {
+							op2 = zend_get_constant_ex(Z_STR_P(op2), ce, ZEND_FETCH_CLASS_SILENT);
+						}
```
This should use zend_update_constant.

pmmaga replied:
Changed. Thanks for the feedback

nikic commented:
Edge-case:

```php
trait T {
    public $x = self::X;
}
trait T2 {
    public $x = self::X;
}
class C {
    use T, T2;
    const X = 42;
}
var_dump((new C)->x);
```

Gives:

```
Warning: Uncaught Error: Cannot access self:: when no class scope is active in /home/nikic/php-src/t084.php:9
Stack trace:
#0 {main}

Next Error: Cannot access self:: when no class scope is active in /home/nikic/php-src/t084.php:9
Stack trace:
#0 {main}
  thrown in /home/nikic/php-src/t084.php on line 9

Fatal error: T and T2 define the same property ($x) in the composition of C. However, the definition differs and is considered incompatible. Class was composed in /home/nikic/php-src/t084.php on line 9
```

pmmaga commented:
@nikic, nice find. I've updated the usage of `zval_update_constant` to `zval_update_constant_ex` which fixes the issue. I've also added your example as a test.

nikic commented:
Slight extension:

```php
<?php

trait T {
    public $x = self::X;
}
trait T2 {
    public $x = self::X;
}
class C {
    use T, T2;
    const X = 42;
}
class D {
    use T2;
    const X = 24;
}
var_dump((new C)->x);
var_dump((new D)->x);
```

This gives 42 42 instead of 42 24. The zend_update_constants_ex shouldn't be performed in-place, but rather on a copy of the zvals.

pmmaga commented:
Updated so that the constant resolution happens on copies.

nikic commented:
Merged as https://github.com/php/php-src/commit/897bdb42f084e20eaf8a4b88829fc94f72626fd0 with a few tweaks (using COPY_OR_DUP and adding dtor calls). Thanks!
